src/ray/object_manager/object_directory.cc

```diff
@@ -216,20 +262,41 @@ ray::Status ObjectDirectory::LookupLocations(const ObjectID &object_id,
                          const std::vector<ObjectTableDataT> &location_history) {
           // Build the set of current locations based on the entries in the log.
           std::unordered_set<ClientID> client_ids;
+          bool inline_object_flag = false;
+          std::vector<uint8_t> inline_object_data;
+          std::string inline_object_metadata;
           bool has_been_created = false;
           UpdateObjectLocations(location_history, gcs_client_->client_table(),
-                                &client_ids, &has_been_created);
+                                &client_ids,
+                                &inline_object_flag,
+                                &inline_object_data,
+                                &inline_object_metadata,
+                                &has_been_created);
           // It is safe to call the callback directly since this is already running
           // in the GCS client's lookup callback stack.
-          callback(object_id, client_ids, has_been_created);
+          callback(object_id, client_ids, inline_object_flag,
+                   inline_object_data, inline_object_metadata,
+                   has_been_created);
         });
   } else {
     // If we have locations cached due to a concurrent SubscribeObjectLocations
     // call, call the callback immediately with the cached locations.
+    // If object inlined, we already have the object's data.
     auto &locations = it->second.current_object_locations;
     bool has_been_created = it->second.has_been_created;
-    io_service_.post([callback, object_id, locations, has_been_created]() {
-      callback(object_id, locations, has_been_created);
+    bool inline_object_flag = it->second.inline_object_flag;
+    std::vector<uint8_t> inline_object_data = it->second.inline_object_data;
```

Comment:
This will do a copy, it might be better to do const auto& inline_object_data = ...